Add deb packaging

keshto opened the pull request:
Address the deb part of https://github.com/open-policy-agent/opa/issues/2210.

keshto commented:
This took longer to figure out the build system then to make the deb package hehe. I would like a little more clarification on this `Add as a step in the release target, document on https://github.com/open-policy-agent/opa/blob/master/docs/devel/RELEASE.md#building`
- It is unclear to me what is to be added to RELEASE.md as I don't think the instructions there would change. I can put a note that addition to all the binaries one would see distro pacakges now as well?
- I don't know if the release targets have changed since this issue was opened, but also some clarity on what is desired here would be great.

patrick-east commented:
> It is unclear to me what is to be added to RELEASE.md as I don't think the instructions there would change. I can put a note that addition to all the binaries one would see distro pacakges now as well?

Well, as-is someone would have to run `make build-distro-packages` as an extra step _or_ we adjust the `make release` step (as documented already) generate them, and then yea, just update the instructions to include the packages in addition to the binaries.

> I don't know if the release targets have changed since this issue was opened, but also some clarity on what is desired here would be great.

I think what you've got right now is good. The only other spot to change would be for `make release` (which I guess really means `./build/build-release.sh`) but we could always do that as a follow up and just document the additional target step for now.

patrick-east commented on build/gen-deb.sh:
```diff
@@ -0,0 +1,45 @@
+cp ../opa_linux_amd64 ./usr/bin/opa
+chmod +x ./usr/bin/opa
+
+tar cJf data.tar.xz ./usr/bin/opa #./usr/share/man/<location> <- in gz format
```
This comment reminds me, can we generate and install the opa man pages with the deb file too?

keshto replied:
done

keshto commented:
> Well, as-is someone would have to run `make build-distro-packages` as an extra step _or_ we adjust the `make release` step (as documented already) generate them, and then yea, just update the instructions to include the packages in addition to the binaries.
> 
Yes, as-is that makes sense, but included if was included into the release I don't see what needs to be changed in the `building` section of the documents. Sorry.
> 
> I think what you've got right now is good. The only other spot to change would be for `make release` (which I guess really means `./build/build-release.sh`) but we could always do that as a follow up and just document the additional target step for now.
That makes a lot more sense. I was trying to understand how an extra step to release target was going to look like. I did play around with adding it to build-release.sh but `xz` is missing from the docker image and I figured that might not be the best way to do it and to ask instead. I will change the docker image and call gen-deb.sh in build-release.sh.

keshto commented:
Added some logic to the gen-deb script to check if we can use dpkg-build instead of manually creating the .deb which works on the docker image without having to change the docker image. I removed the WIP in the title and should be good to go once it has been code reviewed.

patrick-east commented on build/build-release.sh:
```diff
@@ -46,7 +46,7 @@ build_release() {
     GOOS=windows GOARCH=amd64 make build
     mv opa_windows_amd64 opa_windows_amd64.exe
-    mv opa_*_* $OUTPUT_DIR
+    cp opa_*_* $OUTPUT_DIR
```
If we switch to this it means older releases that are missing `build/gen-deb.sh` will leave behind binaries.

If you move the call to `build_distro_packages` to be in this function right after we build the targets it should simplify some of the error cases/cleanup for that.

keshto replied:
For sure, didn't know if it was ok to have everything in one function. I thought that the binary that would be left behind is on the docker image which is ephemeral is that not the case?

keshto replied:
Please see updated changes. Moved the call to `build_distro_packages` into this function as requested.

patrick-east commented on build/gen-deb.sh:
```diff
@@ -0,0 +1,67 @@
+
+md5sum ../opa_linux_amd64 > md5sums
+
+if command -v dpkg-deb >/dev/null 2>/dev/null; then
+  mkdir -p opa_$VERSION/DEBIAN
+  mv control opa_$VERSION/DEBIAN/control
+  mv md5sums opa_$VERSION/DEBIAN/md5sums
+  mv ./usr opa_$VERSION/usr
+
+  dpkg-deb -b opa_$VERSION/
+else # We will do this manually
+  echo 2.0 > debian-binary
+
+  tar cJf data.tar.xz ./usr/bin/opa ./usr/share/man/*
+  tar cJf control.tar.xz control md5sums
+
+  ar r opa_$VERSION.deb debian-binary control.tar.xz data.tar.xz
+fi
```
I'd prefer we only had one path for this, aside from arguably be the "right way" is there any benefit to using `dpkg-deb` over the more manual archive creation?

Would it make sense to just assume we have `dpgk-deb` and have the make targets rely on running in a container which contains the tooling required?

keshto replied:
Assuming `dpkg-deb` is probably the best route. If it tries to get built on a different container/host system that does not have `dpkg-deb` that would be a problem and hence the manual way. However, I think it is ok to reduce complexity in exchange for a simple comment that says we assume `dpkg-deb` to be there.

keshto replied:
Removed the branch and added a check to ensure we have `dpkg-deb` before script continues.

tsandall commented:
hey @keshto thanks for submitting this! Two super high-level questions.

1. Are you planning to use the result of this yourself?

2. Are you able to maintain these changes going forward?

keshto commented:
> hey @keshto thanks for submitting this! Two super high-level questions.
> 
>     1. Are you planning to use the result of this yourself?
> 
>     2. Are you able to maintain these changes going forward?

1. I mostly use Debian based distros for work so I will be using it :)
2. What would need maintaining? I foresee only if the build system changes this would need updating and should be the responsibility of whomever is doing the build system change imho.

patrick-east commented on Makefile:
```diff
@@ -227,6 +241,10 @@ push-binary-edge:
 	aws s3 cp opa_windows_$(GOARCH).exe s3://opa-releases/edge/opa_windows_$(GOARCH).exe
 	aws s3 cp opa_linux_$(GOARCH) s3://opa-releases/edge/opa_linux_$(GOARCH)
 
+.PHONY: push-distro-packages-edge
+push-distro-packages-edge:
+	aws s3 cp opa_$(VERSION).deb s3://opa-releases/edge/opa_$(VERSION).deb
+
```
Lets remove this for now and we can re-add when we integrate with the CI system.

patrick-east commented on Makefile:
```diff
@@ -197,6 +207,10 @@ build-windows:
 	@$(MAKE) build GOOS=windows
 	mv opa_windows_$(GOARCH) opa_windows_$(GOARCH).exe
 
+.PHONY: build-distro-packages
+build-distro-packages:
+	@$(MAKE) deb
+
```
Lets remove this and just keep `deb` target for now. Once we iterate and add more targets we can re-add as needed.

patrick-east commented on build/gen-deb.sh:
```diff
@@ -0,0 +1,63 @@
+echo "Version: $VERSION" >> control
+cat << 'EOF' >> control
+Architecture: amd64
+Maintainer: OPA Team <team@openpolicyagent.org>
```
nit: This email address doesn't exist, lets use cncf-opa-maintainers@lists.cncf.io
